node/src/components/chain_synchronizer/operations.rs

```diff
@@ -894,8 +898,33 @@ async fn fetch_forward(
 async fn fetch_to_genesis(trusted_block: &Block, ctx: &ChainSyncContext<'_>) -> Result<(), Error> {
     let _metric = ScopeTimer::new(&ctx.metrics.chain_sync_fetch_to_genesis_duration_seconds);
 
+    let locally_available_block_range = ctx
+        .effect_builder
+        .get_available_block_range_from_storage()
+        .await;
+
     let mut walkback_block = trusted_block.clone();
     loop {
+        // The available range from storage indicates a range of blocks for which we already have
+        // all the corresponding deploys and global state stored locally.  Skip fetching for such a
+        // range.
+        if locally_available_block_range.contains(walkback_block.height()) {
+            let maybe_lowest_available_block = ctx
+                .effect_builder
+                .get_lowest_available_block_from_storage()
+                .await;
+            if let Some(lowest_available_block) = maybe_lowest_available_block {
+                if lowest_available_block.height() == 0 {
+                    break;
+                }
+                walkback_block = *fetch_and_store_block_by_hash(
+                    *lowest_available_block.header().parent_hash(),
+                    ctx,
+                )
+                .await?;
+            }
+        }
```

Comment:
Do we need to repeat that at the every step of the loop? Once we get the _lowest available block from storage_ its parent will be the next _lowest available block from storage_.

Reply:
We should only actually go inside the `if` block once, as the available range is taken before we enter the loop and it's not updated on each iteration.  Maybe I should rename the variable to `locally_available_block_range` to `available_block_range_on_startup` or something?

Reply:
If that's true then why not extract if outside of the loop? At least it will make it clear to the reader that it will be executed only once.

Reply:
Since Fraser is on holidays this week, I will merge this PR now and put up a following one with the suggested changes.